design/restore-with-EnableAPIGroupVersions-feature.md

```diff
@@ -0,0 +1,193 @@
+# Restore API Group Version by Priority Level When EnableAPIGroupVersions Feature is Set
+
+Status: Draft
+
+## Abstract
+
+This document proposes a solution to select an API group version to restore from the versions backed up using the feature flag EnableAPIGroupVersions.
+
+## Background
+
+It is possible that between the time a backup has been made and a restore occurs that the target Kubernetes version has incremented more than one version. In such a case where multiple versions of Kubernetes were skipped, the preferred source cluster's API group versions for resources may no longer be supported by the target cluster. With PR #2373, all supported API group versions were backed up if the EnableAPIGroupVersions feature flag was set during Velero installation. The next step (outlined by this design proposal) will be to see if any of the backed up versions are supported in the target cluster and if so, choose one to restore for each backed up resource.
+
+## Goals
+
+- Restore resources using the chosen API group version.
+- Choose an API group to restore from backups given a priority system or a user-provided prioritization of versions.
+
+## Non Goals
+
+- Allow users to restore onto a cluster that is is running a Kubernetes version older than the source cluster. The changes proposed here only allow for skipping ahead to a newer Kubernetes version, not older.
+- Allow restoring from backups created using Velero version 1.3 or older.
+- Modifying the compressed backup tarball files.
+- Using plugins to restore a resource when the target supports non of the source cluster's API group versions.
+
+## High-Level Design
+
+During restore, the proposal is that Velero will determine if the `APIGroupVersionsFeatureFlag` was enabled in both the source and target clusters (during backup and restore, respectively). Only if the feature flag was enabled in both clusters will the changes proposed here take effect.
+
+The proposed code starts with creating three lists for each backed up resource. The three lists will be created by
+  (1) reading the directory names in the backup tarball file and seeing which API group versions were backed up from the source cluster,
+  (2) looking at the target cluster and determining which API group versions are supported, and
+  (3) getting config maps on the cluster in order to get user-defined prioritization of versions.
+
+  The three lists will be used to create a map of chosen versions to restore. If there is a user-defined list of priority versions, the versions will be checked against the lists of target and source supported versions. The highest user-defined priority version that is/was supported by both target and source clusters will be the chosen version for that resource.
```

Comment:
Yes, if the user provides a value that doesn't exist in the source and/or target (the value must be in both), then the next paragraph will be executed. The next paragraph is to follow our pre-determined version prioritization (e.g. target preferred version is the highest priority). 

I'll add a sentence that if the user specified version is not supported by both target and source, it will be logged and the restore will continue with other priortizations.

Reply:
Thanks! I found it later in the document too, but I think a brief sentence here would be helpful.